Homogenize the User Interface

jwilm opened the pull request:
Several of the frequently visited views in DCIY have browser default styling which results in a non uniform UI. The current UI is shown below and followed by the UI updates in this PR.
## Unstyled Views

<img src="https://f.cloud.github.com/assets/4285147/1749361/bc33c218-650e-11e3-9f0a-904bd828af96.png" width="288" alt="new_project_path" />

<img src="https://f.cloud.github.com/assets/4285147/1749383/1e5a62c0-6510-11e3-9a2a-76a9ae30c471.png" width="238" alt="projects_path" />

<img src="https://f.cloud.github.com/assets/4285147/1749385/6a6bf034-6510-11e3-9388-a6af26c043cf.png" width="163" alt="new_build_path" />
## UI Updates in this PR

This PR updates those views with styling consistent with the rest of DCIY. A button-danger class is added to support this.

<img src="https://f.cloud.github.com/assets/4285147/1749395/36ac7754-6511-11e3-9d63-38358709e68a.png" width="61" alt="button button-danger" />
### new_project_path

All of the form labels had **bold** set as the `font-weight`. The `https://github.com/` URL snippet was placed in a span so that the input field could have the same font and size without disrupting the rest of the UI. The `<h1>` tag was nested as in other pages to provide a uniform style. The submit input had ".button" class added.
<img src="https://f.cloud.github.com/assets/4285147/1749400/7fe6ccee-6511-11e3-8239-046c7f544e10.png" width="266" alt="new_project_path" />
### projects_path

Here is where most of the effort was input on this PR. As noted above, a `.button-danger` class was added to support a visual cue as to the nature of the `Remove` button. Also note the "Manage Projects" link to this page on the side. The project list was formerly set up as a table. Here we use a `<ul>` to organize the data. The controls are nested in a span which is floated to the right.

You may want to review how the SCSS is set up for the red button style - My CSS-Fu is a bit weak, and I don't have a good field for how to manage variables for a bunch of related but non-identical colors. There's also an issue with the buttons descending below the text on the left.
![projects_path](https://f.cloud.github.com/assets/4285147/1749405/9fc14814-6511-11e3-8135-2dba6e3defaa.png)
### new_build_path

Here the `<h1>` text is nested in the `page-header` and `page-title` divs as on other pages. The "Back" link is now inline with the form actions, and the submit button has the `.button` class.
<img src="https://f.cloud.github.com/assets/4285147/1749407/e429bcfc-6511-11e3-920a-62a966f622e6.png" width="188" alt="new_build_path" />
### build_path

The `<p>` and `<strong>` tags here are replaced with a description list. The lines are slightly separated (vertically and horizontally) to make it feel less cluttered. You will also notice that the verbose "SHA for build.." text is replaced with "SHA". An SHA placeholder was added for when a build is first started - I figured "HEAD" was appropriate since DCIY just lets you specify a repository and a branch.
<img src="https://f.cloud.github.com/assets/4285147/1749413/7edd390e-6512-11e3-9d4c-fd28f91492b2.png" width="473" alt="build_path" />

For completed builds, the commit hash display normally.
<img src="https://f.cloud.github.com/assets/4285147/1749431/3f00d130-6515-11e3-92e8-25cb50f0c1fb.png" width="541" alt="build_path when build is complete" />
## Other comments

I built this changes on top of my last PR which has not yet been merged. If there is a better method for submitting multiple PRs, I would love to read about it.

Cheers!

jwilm committed:
Build meta data uses description list

The format of the Branch/SHA/Status blocks on top of Build#show
suggested a description list would be the semantically correct tags
for displaying that data. Styling is included for a clean look.

jwilm committed:
.build-output margins balanced

The width of .build-output is increased to match the inner with of
div.primary-pane. This results in identical margins on either side
of the build output.

jwilm committed:
Add link to projects_path in main layout

The new-project class is replaced with project-nav to avoid redundant
classes for sidebar nav links.

jwilm committed:
Homogenize project form

- GitHub URL and input for owner/repo share font and size. Font is
  applied using a mixin extracted from the html, body styling.
- Create/Save Project submit now has "button" class.

jwilm committed:
Style projects_path

- Add "button-danger" class for the Remove button
- Changes show anchor text to project.repo
- Removes `table`, use `ul` instead.
- Create _project partial
- Adds project-info class for each list item.

jwilm committed:
Default for build.sha on first run

When first running a build, the SHA is not known if only a branch
is specified. "HEAD" seems like a reasonable default since DCIY
uses the HEAD ref for the branch being built.

jwilm committed:
Style Builds#new view

- Add button class to form submission
- Place 'back' link in .actions div
- Nest h1 within page-header page-title divs

cobyism commented on app/views/builds/new.html.erb:
```diff
@@ -24,8 +28,6 @@
 
   <div class="actions">
-    <%= f.submit %>
+    <%= f.submit class: "button" %> <%= link_to 'Back', builds_path %>
```
:+1: I like having these inline. There’s a pattern I often use for `button + secondary link` that I’d like to follow here and that is  to have the word "or" between the button and the link, and having the secondary link be lowercase so the whole combination reads in sentence case. For example, this whole thing might instead be:

``` erb
<%= f.submit class: "button" %> or <%= link_to 'cancel and go back', builds_path %>
```

Also, the wording "Create Build" isn’t really ideal for this action—something like "Start build" would be more appropriate (or if we wanted to be a little bit fun we could make the button say something like ["Punch it, chewie!"](http://truegif.com/pictures/gif/2731.gif)?). Either way it’d be good to improve the wording used on the button while we’re changing this :grin:

jwilm replied:
My preference would be a Picard [Engage!](http://i2.wp.com/www.iniciativanerd.com.br/wp-content/uploads/2013/05/jean-luc-picard-engage-400x305.jpeg) :smile:. I guess it doesn't have the same impact in button form. Oh well.

Your button pattern would help to give the app some personality. I will add in next commit.

cobyism commented on app/assets/stylesheets/main.scss:
```diff
@@ -37,3 +57,19 @@ li {
+}
+
+dd {
+  margin-left: ($component-width-left + 10px);
```
Do we really need such a big left indent on this and the `dt` above? I think this would probably look better if we only had a little one like this:

![2013-12-21 at 10 47 am](https://f.cloud.github.com/assets/296432/1795650/71927c9e-6a2d-11e3-8680-6989ed671510.png)

What are your thoughts?

cobyism commented:
:sparkling_heart: Thanks for the improvements/contributions man! Great to see! :smiley:

> note the "Manage Projects" link to this page on the side.

I see what you’re trying to do, but I actually think a better course of action with the whole "manage projects" screen is to :fire: it completely. I think it’s only in there in the first place from when I generated a scaffold when I started building this, but between the list of projects in the sidebar, the new project button below, and the "configure" project button/screen, I think all the actions necessary can be done through the existing views.

The only thing that is possible from the "manage projects" screen that isn’t possible from anywhere else is deleting a project, so if we just moved that button to be only shown on the "configure project" page, there would no longer be any need for the standalone project index view:

![2013-12-21 at 10 54 am](https://f.cloud.github.com/assets/296432/1795665/56223b24-6a2e-11e3-9702-ab30f1cfbdd4.png)

If we did that, we could actually remove a whole bunch of code (views, partials, routes, controller actions etc.). Does that make sense to you? What are your thoughts?

Everything else I think I’m pretty happy with. I might have a couple of minor comments about the CSS/class names and stuff (going to have a look over it now), but yeah these improvements will be great to have—thanks a bunch!

cobyism commented on app/assets/stylesheets/_variables.scss:
```diff
@@ -5,6 +5,9 @@ $dciy-red: #911;
 $dciy-grey: #888;
 
+// Sizes
+$component-width-left: 160px;
```
Can we make this variable name a bit more descriptive? `component` is kinda vague, and also maybe the `left` segment might be unnecessary (i.e. the same width might be used as a right margin or something in the future too, depending on how you impelement things).

cobyism commented on app/assets/stylesheets/main.scss:
```diff
@@ -7,14 +7,34 @@
   background: #eee;
 }
 
+div.field {
```
Minor nitpick: the element portion of this selector (`div`) shouldn’t be necessary. Just using the class name is sufficient :neckbeard:

cobyism commented on app/assets/stylesheets/main.scss:
```diff
@@ -7,14 +7,34 @@
+    font-weight: bold;
+  }
+
+  span {
```
The `label` and `input` selectors are clear enough given their function, but can we maybe give this `span` an intention-revealing class name and use that as the CSS selector instead? It’d be great to know in the future just by looking at the CSS what is actually being styled here, but at the moment I have no idea where/how this span (or how many of them) are part of the component/markup.

cobyism commented on app/views/builds/show.html.erb:
```diff
@@ -4,21 +4,25 @@
+
+
+  <dt>SHA</dt>
+  <dd><%= @build.sha || "HEAD" %></dd>
```
I’m happy to just show "HEAD" if the sha isn’t there yet, but would it be better to just not show it here at all until it’s determined? Can we hook this up to the JS auto-update stuff so that it turns up once it has been computed? Not a major problem by any means, but just worth seeing if there’s something easy we can do here to improve things.

cobyism commented:
> I built this changes on top of my last PR which has not yet been merged. If there is a better method for submitting multiple PRs, I would love to read about it.

This shouldn’t be a problem, but in the future you shouldn’t have to worry about this. The two PRs alter completely different sets of files, so they should both be able to be worked on, submitted, and merged independently. It’s not really that obvious, but in future, the best way to keep PRs independent would be to:
- create a branch for the first change from `master` (or whatever the primary branch is for the project).
- commit those changes, push them up, and send a PR for them.
- go back and check out `master`
- create a second branch from `master` for the next set of changes
- commit those changes push them up, send a PR etc.
- repeat for as many changes as you want to make.

Even if changes you make do end up conflicting with each other (such as two alternative implementations for the same thing), this is usually the best approach since it means that everything stays related and self-contained in each branch, and means that the person evaluating whether to merge one PR doesn’t have to worry about whether changes from other branches are included or not. For example, if (hypothetically) I was opposed to the changes you made in #22 but wanted to merge in these changes, it would be much more complex (although still possible by rewriting history in git) to make that happen since this branch is built on top of those commits. There shouldn’t be any problem here, but if there is we’ll sort it out. Thanks for bringing that point up! :grinning:

jwilm commented:
Having a separate "manage projects" page creates visual redundancy. The list of projects on the right is present in all the views, so if we add a `Remove Project` button to the `edit_project` path then we can do away with this view and a good amount of code as you suggest.

Descriptive naming in stylesheets isn't particularly natural for me, so I appreciate the comments about the CSS here. I will get that cleaned up.

The `<dt>` tag width started out at that size since the "SHA" label used to be quite a bit longer, and narrowing that down now makes a lot of sense.

jwilm committed:
Resize description title and rename sass variable

Description list description title is reduced and now measured in
`em`.

jwilm committed:
Remove index action from Projects

- Scrub view/js builder
- Remove index action
- Update router
- Remove link to projectspath

jwilm committed:
Build creator funification

- Change text on build runner
- Go back button now returns user to the project_builds path for
  the current project

jwilm committed:
Add destroy link to edit project

- Destroy button now on project edit view
- Redirect to root path after project destroy
- Add intelligent go back button to project form partial

jwilm commented:
I played around with eliminating the `edit_project` path and moving the destroy button to the top of the `project_builds` page. It looked ok (below), but it didn't make sense having that button near other much more common actions, and we would've had to bring the edit route back when we add more configurables (post build hooks, default branch, etc).

<img src="https://f.cloud.github.com/assets/4285147/1798258/ce5a1226-6b3a-11e3-8309-50ec99c60c80.png" width="777"/>

All that is scrapped, and the destroy button is presently placed below the form actions on the `edit_project` page.

<img src="https://f.cloud.github.com/assets/4285147/1798329/30e08612-6b42-11e3-98dd-68dfa07a7f20.png" width="281" />

I just noticed while typing this that there's a minor button size issue when the text has both ascenders and descenders. Note how the "go back" button is slightly shorter than its neighbor.

Unless I missed something these should be the remaining actions for this PR:
- CSS selectors should be more descriptive
- Maybe remove `SHA` field from show project until subsequent views. Hooking up a javascript update would be trivial as you suggest. Personally, I just like knowing that it is building the latest commit on branch _x_, so I am inclined to leave it saying "**SHA** HEAD". I guess that's the default behavior though, so we could just replace HEAD with the actual hash once we know it.

Thanks for the comments on managing independent PRs. This was sort of the conclusion I came to after submitting this PR, and I appreciate the validation :grin:.

jwilm committed:
Update redirects in specs, fix Build#new

Redirects for some UI actions were made more user friendly in the
last commit, and the specs are now updated to match. Build#new
had an issue when visiting builds_path (instead of
project_builds_path). Both situations are now function as expected,
and all specs pass.

jwilm committed:
Build updater script only runs on Build#show

Adds a check before kicking off interval update script that window
path looks like /builds\/\d+.

jwilm committed:
Add JavaScript update of SHA during build process

Adds a span.js-build-sha within the SHA <dd> tag on Build#show. The
javascript updater checks if it has been set and updates from
build.sha. The `sha` attribute is now served from the build#show
jsbuilder.

jwilm committed:
Project now set correctly on project_builds_path

Set project on @build instead of a separate instance variable. This
reduced overall code and fixed problem with setting which project
to build.

jwilm committed:
Clean up CSS and Project#index leftovers

.field span now selected using .field .description instead. Some
styling added/changed before canning Project#index was reverted.
Unused partial was removed.

jwilm commented:
This should all be good to go now. I realized while working on this that the error produced when not having a project selected when creating a new build needs some style help, too. I think that might be better off in another PR though.

I liked your "punch it, chewie!" idea - here's the updated new build page.
<img src="https://f.cloud.github.com/assets/4285147/1818697/9f0e1bd8-701d-11e3-9d73-3bb535391aab.png" width="316" />

If everything is good to go, I can rebase this on top of master to omit that other feature branch. Sorry for the confusion there.

parkr commented:
Would love to see this added. :+1:

cobyism commented:
Rad, yeah if you could rebase or fix the conflicts with master I’ll take another last review and then merge in :+1: :sparkles:

jwilm commented:
This was rebased on top of the latest _master_ last night. The commits in #22 have been omitted.

cobyism commented:
Awesome, I’ll try and take a look at it soon :grinning:

cobyism commented:
This `<hr>` needs to be given some styles (probably an even top/bottom margin, and just make it a single light-grey line, or something):

![2014-01-08 at 4 44 pm](https://f.cloud.github.com/assets/296432/1870197/31758a06-7884-11e3-99fc-a62f19fefc7c.png)

From screenshots above in the thread, it looks like it _was_ styled at some point, but maybe those got taken out somewhere? Also, with three buttons like this all so close together, I feel like the button to delete the project should be made less prominent (maybe try adding a `.small` class to it), so that people don’t mistake it for being the "primary" action on this screen.

Tiny nitpick: the text shadow on the danger button should probably be changed so that it’s _darker_ than the background instead of lighter:

![2014-01-08 at 4 50 pm](https://f.cloud.github.com/assets/296432/1870278/2c3b43e0-7885-11e3-8703-48d29dee5f11.png)

Having the text-shadow be lighter than the background on grey buttons makes sense, because there the text color is darker than the background. Anytime I have buttons where the button color is _darker_ than the text though (i.e red background is darker than white text), I think it makes more sense to have the shadow be slightly darker than the button background (either `-1px` for an inset text effect, or just regular `1px` to have the text appear raised off the button works too).

> I just noticed while typing this that there's a minor button size issue when the text has both ascenders and descenders. Note how the "go back" button is slightly shorter than its neighbor.

I wouldn’t worry about that—I’m pretty sure that’s just due to the fact that the underlying element being styled as `.button` is an `input[type="button"]` in one case and a regular `a` in the other, and my rudimentary button CSS hasn’t made them identical.

Other than that, I think this PR is good to go. There’s a few small things I noticed while reviewing, but they’re problems in `master` anyway so nothing to do with this PR :grin:

jwilm commented:
Ooops, I must have removed that `<hr>` style when I trashed the _Manage Projects_ page. :flushed:

CSS from old commit:

``` css
hr {
   background-color: #ccc;
   height: 1px;
   border: 0px;
   color: #ccc;
   margin: 1em 0;
 }
```

I'll add that back in later. The effect was a muted 1px line.

I like how your dark shadow on the button makes it appear more crisp. Interestingly, mixing inset and raised effects for dark and light text seems to retain a consistent appearance (judging by the buttons below this comment form.. which I am starting to suspect you had something to do with!). 

Your `.small` suggestion for that button makes a lot of sense. Since you brought up the issue of highlighting the _primary_ action on a page, maybe we should add a `.primary` button class. I'm thinking green for "go" given that running a build is probably the most common action.

Thanks for all the attention to detail and for your feedback! :heart:

jwilm committed:
Fix style on edit project page

- Reintroduce <hr> style
- "Remove Project" button is .small
- dark shadow on .button-danger text

jwilm committed:
Style .button-primary for new build

- Add &.button-primary in .button
- Add .button-primary class to primary button on Builds#new.
- Fix formatting in project _form HTML

jwilm commented:
The last two commits address the latest discussion in this thread. From the commits:
- Reintroduce `<hr>` style
- "Remove Project" button is `.small`
- dark shadow on `.button-danger` text
- Add &.button-primary in .button
- Add `.button-primary` class to primary button on Builds#new.
- Fix formatting in project _form HTML

Here are the updated screenshots.

### Edit Project

<img src="https://f.cloud.github.com/assets/4285147/1889183/aa7f8426-7a17-11e3-81c2-ecd5261e780d.png" width="278" />

I didn't add the .button-primary class here since it's right next to a danger button and started to look like christmas, but semantically speaking it should probably be different from the `cancel` button.

### New Build

<img src="https://f.cloud.github.com/assets/4285147/1889187/c25a9dd8-7a17-11e3-8d0c-d7ddf7dc8606.png" width="335" />

Everything builds locally (cannot rely on DCIY output as per #32), and I've looked around at all of the pages for a consistent style. This is rebased on top of the current _master_, and I _think_ we should be good to go!

cobyism commented:
:sparkles: Cool, sounds good to me! There’s still a long way to go in the UI for this project, but this PR has made a good dent in at least having it be more consistent. Thanks a bunch @jwilm!
